### PR TITLE
CLN: 29547 replace old string formatting

### LINKED ISSUE #29547: Replace old string formatting syntax with f-strings
Since we no longer support python 3.5, we can now use the new f-strings instead of the old ```.format()``` ( and obviously the ```%``` formatting).

#### Notes:
* Don't forget to link this issue in your pull request's body message , simply paste this ```https://github.com/pandas-dev/pandas/issues/29547``` in your pull request's body message.

* If any of your changed files are related to #29886 , please make sure to link your pull request to that issue as well, simply paste this ```https://github.com/pandas-dev/pandas/issues/29886``` in your pull request's body message as well.

* Please comment what you are planning to work on, so we won't do double work.

* If a file/files that should be marked as done, is not marked, please comment letting me know.

---

To check what files still needs to be fixed in the ```pandas``` directory:
```
grep -l -R '%s'  --include=*.{py,pyx} pandas/
grep -l -R '%d' --include=*.{py,pyx} pandas/
grep -l -R '\.format(' --include=*.{py,pyx} pandas/
```

All of the above can also be used as a one liner:
```
grep -l -R -e '%s' -e '%d' -e '\.format(' --include=*.{py,pyx} pandas/
```

##### Tip:
If you want to see the line number of the occurrence, replace the ```-l``` with ```-n```
for example:
```
grep -n -R '%s' --include=*.{py,pyx} pandas/
```

---

### The current list is:

- [ ] pandas/compat/pickle_compat.py

- [ ] pandas/_config/config.py

- [ ] pandas/core/arrays/datetimelike.py

- [ ] pandas/core/arrays/datetimes.py

- [ ] pandas/core/arrays/integer.py

- [ ] pandas/core/arrays/period.py

- [ ] pandas/core/computation/pytables.py

- [ ] pandas/core/config_init.py

- [ ] pandas/core/frame.py

- [ ] pandas/core/generic.py

- [ ] pandas/core/groupby/generic.py

- [ ] pandas/core/groupby/groupby.py

- [ ] pandas/core/indexes/base.py

- [ ] pandas/core/indexes/multi.py

- [ ] pandas/core/indexes/range.py

- [ ] pandas/core/ops/docstrings.py

- [ ] pandas/core/ops/\_\_init__.py

- [ ] pandas/core/reshape/merge.py

- [ ] pandas/core/tools/datetimes.py

- [ ] pandas/io/formats/css.py

- [ ] pandas/io/formats/excel.py

- [ ] pandas/io/formats/format.py

- [ ] pandas/io/formats/html.py

- [ ] pandas/io/formats/info.py

- [ ] pandas/io/formats/latex.py

- [ ] pandas/io/formats/printing.py

- [ ] pandas/io/formats/style.py

- [ ] pandas/io/parsers.py

- [ ] pandas/io/pytables.py

- [ ] pandas/io/sas/sas_xport.py

- [ ] pandas/io/stata.py

- [ ] pandas/_libs/tslibs/c_timestamp.pyx

- [ ] pandas/_libs/tslibs/frequencies.pyx

- [ ] pandas/_libs/tslibs/parsing.pyx

- [ ] pandas/_libs/tslibs/period.pyx

- [ ] pandas/_libs/tslibs/strptime.pyx

- [ ] pandas/_libs/tslibs/timedeltas.pyx

- [ ] pandas/plotting/_matplotlib/converter.py

- [ ] pandas/tests/arrays/categorical/test_operators.py

- [ ] pandas/tests/arrays/test_datetimelike.py

- [ ] pandas/tests/dtypes/test_dtypes.py

- [ ] pandas/tests/extension/base/setitem.py

- [ ] pandas/tests/frame/test_constructors.py

- [ ] pandas/tests/frame/test_missing.py

- [ ] pandas/tests/frame/test_to_csv.py

- [ ] pandas/tests/groupby/aggregate/test_other.py

- [ ] pandas/tests/indexes/datetimes/test_date_range.py

- [ ] pandas/tests/indexes/datetimes/test_datetime.py

- [ ] pandas/tests/indexes/datetimes/test_formats.py

- [ ] pandas/tests/indexes/datetimes/test_partial_slicing.py

- [ ] pandas/tests/indexes/interval/test_constructors.py

- [ ] pandas/tests/indexes/interval/test_interval.py

- [ ] pandas/tests/indexes/multi/test_format.py

- [ ] pandas/tests/indexes/period/test_formats.py

- [ ] pandas/tests/indexes/test_base.py

- [ ] pandas/tests/indexes/timedeltas/test_timedelta.py

- [ ] pandas/tests/indexing/test_categorical.py

- [ ] pandas/tests/indexing/test_coercion.py

- [ ] pandas/tests/io/excel/test_openpyxl.py

- [ ] pandas/tests/io/excel/test_writers.py

- [ ] pandas/tests/io/formats/test_format.py

- [ ] pandas/tests/io/formats/test_printing.py

- [ ] pandas/tests/io/formats/test_style.py

- [ ] pandas/tests/io/formats/test_to_csv.py

- [ ] pandas/tests/io/formats/test_to_html.py

- [ ] pandas/tests/io/formats/test_to_latex.py

- [ ] pandas/tests/io/parser/test_compression.py

- [ ] pandas/tests/io/parser/test_encoding.py

- [ ] pandas/tests/io/parser/test_header.py

- [ ] pandas/tests/io/parser/test_parse_dates.py

- [ ] pandas/tests/io/parser/test_usecols.py

- [ ] pandas/tests/io/test_html.py

- [ ] pandas/tests/io/test_sql.py

- [ ] pandas/tests/io/test_stata.py

- [ ] pandas/tests/reductions/test_reductions.py

- [ ] pandas/tests/reshape/test_concat.py

- [ ] pandas/tests/reshape/test_melt.py

- [ ] pandas/tests/scalar/period/test_period.py

- [ ] pandas/tests/scalar/timedelta/test_timedelta.py

- [ ] pandas/tests/scalar/timestamp/test_constructors.py

- [ ] pandas/tests/series/indexing/test_numeric.py

- [ ] pandas/tests/series/indexing/test_take.py

- [ ] pandas/tests/series/indexing/test_where.py

- [ ] pandas/tests/series/methods/test_rename.py

- [ ] pandas/tests/series/test_api.py

- [ ] pandas/tests/series/test_constructors.py

- [ ] pandas/tests/series/test_datetime_values.py

- [ ] pandas/tests/series/test_repr.py

- [ ] pandas/tests/test_strings.py

- [ ] pandas/tests/tools/test_to_datetime.py

- [ ] pandas/tests/tseries/holiday/test_calendar.py

- [ ] pandas/tests/tseries/holiday/test_holiday.py

- [ ] pandas/tests/tslibs/test_parsing.py

- [ ] pandas/tests/util/test_assert_frame_equal.py

- [ ] pandas/tseries/frequencies.py

- [ ] pandas/util/_decorators.py

- [ ] pandas/util/_test_decorators.py

- [ ] pandas/util/_validators.py

- [ ] pandas/_version.py

---

#### NOTE:

The list may change as files are moved/renamed constantly.

---

Inhereted files and commands from this [PR](https://github.com/pandas-dev/pandas/issues/16130).

### PR DESCRIPTION
**Pull request by 3vts**
I splitted PR #31844 in batches, this is the **last**
For this PR I ran the command `grep -l -R -e '%s' -e '%d' -e '\.format(' --include=*.{py,pyx} pandas/` and checked all the files that were returned for `.format(` and changed the old string format for the corresponding `fstrings` to attempt a full clean of, [#29547](https://github.com/pandas-dev/pandas/issues/29547). I may have missed something so is a good idea to double check just in case

- [ x  ] tests added / passed
- [ x ] passes `black pandas`
- [ x ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ x ] Ref #29547

### CODE REVIEW

File: pandas/io/formats/format.py [hunk 1]
```diff
@@ -217,7 +217,7 @@ def to_string(self) -> str:
 
         fmt_values = self._get_formatted_values()
 
-        fmt_values = ["{i}".format(i=i) for i in fmt_values]
+        fmt_values = [f"{i}" for i in fmt_values]
```
Comment (simonjayhawkins) [inline pandas/io/formats/format.py]:
```suggestion
        fmt_values = [str(i) for i in fmt_values]
```

although should be able to remove line entirely since _get_formatted_values on line above returns List[str]

File: pandas/io/formats/format.py [hunk 2]
```diff
@@ -1257,17 +1245,17 @@ def _format(x):
...
-                fmt_values.append(" {v}".format(v=_format(v)))
+                fmt_values.append(f" {_format(v)}")
             elif is_float_type[i]:
                 fmt_values.append(float_format(v))
             else:
                 if leading_space is False:
                     # False specifically, so that the default is
                     # to include a space if we get here.
-                    tpl = "{v}"
+                    tpl = f"{_format(v)}"
```
Comment (simonjayhawkins) [inline pandas/io/formats/format.py]: this is a template so I _think_ better to leave as is. otherwise change the variable name.

File: pandas/io/formats/format.py [hunk 3]
```diff
@@ -1716,7 +1704,7 @@ def _formatter(x):
             x = Timedelta(x)
         result = x._repr_base(format=format)
         if box:
-            result = "'{res}'".format(res=result)
+            result = f"'{result}'"
```
Comment (simonjayhawkins) [inline pandas/io/formats/format.py]: I _think_ this may be equivalent to
```suggestion
            result = repr(result)
```

File: pandas/io/formats/html.py [hunk 4]
```diff
@@ -149,16 +149,13 @@ def _write_cell(
...
             end_a = "</a>"
         else:
             end_a = ""
 
         self.write(
-            "{start}{rs}{end_a}</{kind}>".format(
-                start=start_tag, rs=rs, end_a=end_a, kind=kind
-            ),
-            indent,
+            f"{start_tag}{rs}{end_a}</{kind}>", indent,
```
Comment (simonjayhawkins) [inline pandas/io/formats/html.py]:
```suggestion
            f"{start_tag}{rs}{end_a}</{kind}>", indent
```

File: pandas/io/formats/latex.py [hunk 5]
```diff
@@ -58,10 +58,9 @@ def write_result(self, buf: IO[str]) -> None:
         """
         # string representation of the columns
         if len(self.frame.columns) == 0 or len(self.frame.index) == 0:
-            info_line = "Empty {name}\nColumns: {col}\nIndex: {idx}".format(
-                name=type(self.frame).__name__,
-                col=self.frame.columns,
-                idx=self.frame.index,
+            info_line = (
+                f"Empty {type(self.frame).__name__}\n"
+                f"Columns: {self.frame.columns}\nIndex: {self.frame.index}"
```
Comment (simonjayhawkins) [inline pandas/io/formats/latex.py]: can you break this at the \n for readability

File: pandas/io/formats/latex.py [hunk 6]
```diff
@@ -140,8 +139,8 @@ def pad_empties(x):
                     buf.write("\\endhead\n")
                     buf.write("\\midrule\n")
                     buf.write(
-                        "\\multicolumn{{{n}}}{{r}}{{{{Continued on next "
-                        "page}}}} \\\\\n".format(n=len(row))
+                        f"\\multicolumn{{{len(row)}}}{{r}}"
+                        "{{Continued on next page}} \\\\\n"
```
Comment (simonjayhawkins) [inline pandas/io/formats/latex.py]: why do two sets of braces need to be removed?

File: pandas/io/formats/printing.py [hunk 7]
```diff
@@ -229,7 +229,7 @@ def as_escaped_string(
             max_seq_items=max_seq_items,
         )
     elif isinstance(thing, str) and quote_strings:
-        result = "'{thing}'".format(thing=as_escaped_string(thing))
+        result = f"'{as_escaped_string(thing)}'"
```
Comment (simonjayhawkins) [inline pandas/io/formats/printing.py]: is this the same as 
```suggestion
        result = repr(as_escaped_string(thing))
```

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @3vts for the PR. a few comments/queries

File: pandas/io/formats/format.py [hunk 8]
```diff
@@ -301,28 +301,26 @@ def _get_footer(self) -> str:
...
 
         if self.name is not False and name is not None:
             if footer:
                 footer += ", "
 
             series_name = pprint_thing(name, escape_chars=("\t", "\r", "\n"))
-            footer += (
-                ("Name: {sname}".format(sname=series_name)) if name is not None else ""
-            )
+            footer += f"Name: {series_name}" if name is not None else ""
```
[the remark this answers is not in the document]
Comment (3vts) [reply pandas/io/formats/format.py]: You are right. In this regard I noticed a lot of places in the code that use `is not None` which AFAIK is redundant. I was thinking on a cleanup for those. What do you think?

File: pandas/io/formats/latex.py (hunk 6 above)
Comment (3vts) [reply pandas/io/formats/latex.py]: This actually got me as well. 2 sets of braces are needed for the final string, since I remove the `.format` and this becomes a regular string there is no need for the "escape" sequence of 2 braces to show 1. Is that or 4 braces with f string

File: pandas/io/formats/format.py (hunk 3 above)
Comment (3vts) [reply pandas/io/formats/format.py]: I think no because the extra single quotes are needed

File: pandas/io/formats/printing.py (hunk 7 above)
Comment (3vts) [reply pandas/io/formats/printing.py]: Not sure because a pair of single quotes is needed

File: pandas/io/formats/latex.py (hunk 5 above)
Comment (3vts) [reply pandas/io/formats/latex.py]: Sure

File: pandas/io/formats/format.py [hunk 9]
```diff
@@ -1238,10 +1226,10 @@ def _format(x):
                     pass
                 return self.na_rep
             elif isinstance(x, PandasObject):
-                return "{x}".format(x=x)
+                return f"{x}"
             else:
                 # object dtype
-                return "{x}".format(x=formatter(x))
+                return f"{formatter(x)}"
```
[the remark this answers is not in the document]
Comment (3vts) [reply pandas/io/formats/format.py]: I am not super comfortable but will give it a try (always good to practice)

File: pandas/io/formats/format.py (hunk 3 above)
Comment (MomIsBestFriend) [reply pandas/io/formats/format.py]: Oh that's okay, the default way ```repr``` works is to wrap the string in extra quotes for example:

```python
txt = "foo"

str(txt) # foo
repr(txt) # 'foo'
```

File: pandas/io/formats/format.py (hunk 1 above)
Comment (jreback) [reply pandas/io/formats/format.py]: yep, let's see if we can remove this entirely

File: pandas/io/formats/format.py (hunk 3 above)
Comment (3vts) [reply pandas/io/formats/format.py]: TIL

File: pandas/io/formats/format.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply pandas/io/formats/format.py]: code cleanups are always welcome.

**Comment by simonjayhawkins** [comment]
@3vts thanks for making the changes. lgtm. can you run black to get ci passing

**Review by simonjayhawkins — APPROVED** [review]
Thanks @3vts lgtm pending green

File: pandas/core/arrays/interval.py [hunk 10]
```diff
@@ -195,13 +195,13 @@ def _simple_new(
         left = ensure_index(left, copy=copy)
         right = ensure_index(right, copy=copy)
 
-        if dtype is not None:
+        if dtype:
```
Comment (MomIsBestFriend) [inline pandas/core/arrays/interval.py]: This doesn't seems right, are you sure about this (I could be 

(Also appears in other spots)

**Comment by MomIsBestFriend** [comment]
> I splitted PR #31844 in batches, this is the **last**
> For this PR I ran the command `grep -l -R -e '%s' -e '%d' -e '\.format(' --include=*.{py,pyx} pandas/` and checked all the files that were returned for `.format(` and changed the old string format for the corresponding `fstrings` to attempt a full clean of, [#29547](https://github.com/pandas-dev/pandas/issues/29547). I may have missed something so is a good idea to double check just in case
> 
>     * [ x  ] tests added / passed
> 
>     * [ x ] passes `black pandas`
> 
>     * [ x ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
> 
>     * [ x ] closes #29547 (?)

@3vts Can you please change the last line from ```Closes #29547``` to ```Ref #29547```, Just so it won't close the master issue

File: pandas/core/arrays/interval.py (hunk 10 above)
Comment (MomIsBestFriend) [reply pandas/core/arrays/interval.py]: Apologizes, there is a ```not``` in there, my bad.

Comment (3vts) [reply pandas/core/arrays/interval.py]: Yep, that is why. I will probably hunt those in the mater repo because with some benchmarks I did I found there is a hair of performance improvement when you remove the `is not None`

Comment (simonjayhawkins) [reply pandas/core/arrays/interval.py]: can you revert the last commit. and do this in a separate PR.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/interval.py (hunk 10 above)
Comment (3vts) [reply pandas/core/arrays/interval.py]: Sure

File: pandas/io/formats/format.py [hunk 11]
```diff
@@ -301,28 +300,26 @@ def _get_footer(self) -> str:
...
-            footer += "Length: {length}".format(length=len(self.series))
+            footer += f"Length: {len(self.series)}"
 
         if self.dtype is not False and self.dtype is not None:
             name = getattr(self.tr_series.dtype, "name", None)
             if name:
                 if footer:
                     footer += ", "
-                footer += "dtype: {typ}".format(typ=pprint_thing(name))
+                footer += f"dtype: {pprint_thing(name)}"
```
Comment (simonjayhawkins) [inline pandas/io/formats/format.py]: will your updating this section, can we rename `name` here to `dtype_name`

**Review by simonjayhawkins — COMMENTED** [review]
Thanks for the revert of the None stuff, but why the revert of the repr() changes?

File: pandas/io/formats/format.py (hunk 11 above)
Comment (3vts) [reply pandas/io/formats/format.py]: I reverted the `repr()` basically because for some reason with that change the `Build Documentation` check fails

Comment (simonjayhawkins) [reply pandas/io/formats/format.py]: I _think_ #32060 was responsible for those ci failures. should be fixed now.

can you revert https://github.com/pandas-dev/pandas/pull/32034/commits/226b72d8f4fcb598275b5b0c0b2fa9a82fcbdae3 and see  if the tests succeed.

Comment (3vts) [reply pandas/io/formats/format.py]: Seems like that's the case. Let's see.....

**Review by simonjayhawkins — APPROVED** [review]
Thanks @3vts lgtm. @MomIsBestFriend

**Review by ShaharNaveh — APPROVED** [review]
LGTM

File: pandas/io/formats/format.py [hunk 12]
```diff
@@ -1716,7 +1703,7 @@ def _formatter(x):
             x = Timedelta(x)
         result = x._repr_base(format=format)
         if box:
-            result = "'{res}'".format(res=result)
+            result = repr(result)
```
Comment (WillAyd) [inline pandas/io/formats/format.py]: Maybe lost in a comment but why was this changed to `repr`?

File: pandas/io/formats/printing.py [hunk 13]
```diff
@@ -229,7 +229,7 @@ def as_escaped_string(
             max_seq_items=max_seq_items,
         )
     elif isinstance(thing, str) and quote_strings:
-        result = "'{thing}'".format(thing=as_escaped_string(thing))
+        result = repr(as_escaped_string(thing))
```
Comment (WillAyd) [inline pandas/io/formats/printing.py]: Same question

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/formats/format.py (hunk 3 above)
Comment (3vts) [reply pandas/io/formats/format.py]: @WillAyd this thread may explain the reasons of the change

Comment (WillAyd) [reply pandas/io/formats/format.py]: We should use the quotes as done originally; this isn't immediately clear to other readers that we would choose `repr` just to have free quoting and is just confusing.

**Review by WillAyd — CHANGES_REQUESTED** [review]
See comment on repr changes - should just use quotes instead

File: pandas/io/formats/format.py [hunk 14]
```diff
@@ -301,28 +300,26 @@ def _get_footer(self) -> str:
...
 
         if self.name is not False and name is not None:
             if footer:
                 footer += ", "
 
             series_name = pprint_thing(name, escape_chars=("\t", "\r", "\n"))
-            footer += (
-                ("Name: {sname}".format(sname=series_name)) if name is not None else ""
-            )
+            footer += f"Name: {series_name}"
```
Comment (WillAyd) [inline pandas/io/formats/format.py]: I don't think this is totally equivalent?

```python
>>> series_name = None
>>> f"Name: {series_name}"
'Name: None'
```

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/formats/format.py (hunk 14 above)
Comment (3vts) [reply pandas/io/formats/format.py]: Take a look to line 305

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @3vts 

General comment - would be nice to keep PRs more minimal in the future. This was maybe an exception because of all of the double curly braces at play, but if something else comes up the same in the future should tackle in smaller bits
